src/mame/machine/tsconf.cpp

```diff
@@ -0,0 +1,622 @@
+	}
+	else if (nreg == REGNUM(PAGE1))
+	{
+		m_bank2->set_base(m_ram->pointer() + PAGE(PAGE1));
+	}
+	else if (nreg == REGNUM(PAGE2))
+	{
+		m_bank3->set_base(m_ram->pointer() + PAGE(PAGE2));
+	}
+	else if (nreg == REGNUM(PAGE3))
+	{
+		m_bank4->set_base(m_ram->pointer() + PAGE(PAGE3));
+	}
+	else if (nreg == REGNUM(BORDER))
+	{
+		m_port_fe_data = data;
+	}
+	else if (nreg == REGNUM(DMAS_ADDRESS_L))
+	{
+		m_dma->set_saddr_l(data);
+	}
+	else if (nreg == REGNUM(DMAS_ADDRESS_H))
+	{
+		m_dma->set_saddr_h(data);
+	}
+	else if (nreg == REGNUM(DMAS_ADDRESS_X))
+	{
+		m_dma->set_saddr_x(data);
+	}
+	else if (nreg == REGNUM(DMAD_ADDRESS_L))
+	{
+		m_dma->set_daddr_l(data);
+	}
+	else if (nreg == REGNUM(DMAD_ADDRESS_H))
+	{
+		m_dma->set_daddr_h(data);
+	}
+	else if (nreg == REGNUM(DMAD_ADDRESS_X))
+	{
+		m_dma->set_daddr_x(data);
+	}
+	else if (nreg == REGNUM(DMA_LEN))
+	{
+		m_dma->set_block_len(data);
+	}
+	else if (nreg == REGNUM(DMA_NUM_L))
+	{
+		m_dma->set_block_num_l(data);
+	}
+	else if (nreg == REGNUM(DMA_NUM_H))
+	{
+		m_dma->set_block_num_h(data);
+	}
+	else if (nreg == REGNUM(DMA_CTRL))
+	{
+		m_dma->start_tx(((BIT(data, 7) << 3) | (data & 0x07)), BIT(data, 5), BIT(data, 4), BIT(data, 3));
+	}
+	else if (nreg == REGNUM(SYS_CONFIG))
+	{
+		/* TODO
+		switch (data & 0x03)
+		{
+		case 2:
+			m_maincpu->set_clock(X1);
+			break;
+		case 1:
+			m_maincpu->set_clock(X1 / 2);
+			break;
+		case 0:
+		default:
+			m_maincpu->set_clock(X1 / 4);
+			break;
+		}
+		*/
+	}
+	else if (nreg == REGNUM(HS_INT) || nreg == REGNUM(VS_INT_L) || nreg == REGNUM(VS_INT_H) || nreg == REGNUM(INT_MASK))
+	{
+	}
+	else
+	{
+		LOGERROR("Unsupported reg write: %02X = %02x\n", nreg, data);
+	}
+}
+
+u8 tsconf_state::tsconf_port_f7_cmos_r(offs_t offset)
+{
+	// BFF7
+	u8 data;
+	if (m_gluk_ext == PS2KEYBOARDS_LOG && m_gluk_reg == 0xf0)
+	{
+		data = m_keyboard->read();
+	}
+	else if (m_gluk_ext != DISABLED)
+	{
+		data = m_cmos->read(m_gluk_reg);
+	}
+	else
+	{
+		data = 0xff;
+	}
+	return data;
+}
+
+void tsconf_state::tsconf_port_f7_cmos_w(offs_t offset, u8 data)
+{
+	auto m_l = offset >> 12;
+	if (m_l == 6) // EF
+	{
+		m_gluk_ext = (data & 0x80) ? CONF_VERSION : DISABLED;
+	}
+	else if (m_gluk_ext != DISABLED)
+	{
+		if (m_l == 5) // DF
+		{
+			// 0x0E..0xEF
+			m_gluk_reg = data;
+			if (m_gluk_reg == 0x00 || m_gluk_reg == 0x02 || m_gluk_reg == 0x04 || m_gluk_reg == 0x06 || m_gluk_reg == 0x07 || m_gluk_reg == 0x08 || m_gluk_reg == 0x09)
+			{
+				system_time curtime;
+				machine().current_datetime(curtime);
+				m_cmos->write(0x00, dec_2_bcd(curtime.local_time.second));
+				m_cmos->write(0x02, dec_2_bcd(curtime.local_time.minute));
+				m_cmos->write(0x04, dec_2_bcd(curtime.local_time.hour));
+				m_cmos->write(0x06, dec_2_bcd(curtime.local_time.weekday));
+				m_cmos->write(0x07, dec_2_bcd(curtime.local_time.mday));
+				m_cmos->write(0x08, dec_2_bcd(curtime.local_time.month));
+				m_cmos->write(0x09, dec_2_bcd(curtime.local_time.year % 100));
+			}
```

Comment:
You should be using a device that uses the `device_rtc_interface` mixin to implement a real-time clock device.  Ad-hoc implementations are problematic.